tests/unit/ViolationUtilsTest.js

```diff
@@ -0,0 +1,192 @@
+        });
+
+        it('should not add a categoryOutOfPolicy violation when categories are not required', () => {
+            policyRequiresCategories = false;
+            transaction.category = 'Drinks';
+
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+            expect(result.value).not.toContainEqual([categoryOutOfPolicyViolation]);
+        });
+
+        it('should add categoryOutOfPolicy violation to existing violations if they exist', () => {
+            transaction.category = 'Bananas';
+            transactionViolations = [
+                {name: 'duplicatedTransaction', type: 'violation', userMessage: ''},
+                {name: 'receiptRequired', type: 'violation', userMessage: ''},
+            ];
+
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+            expect(result.value).toEqual(expect.arrayContaining([categoryOutOfPolicyViolation, ...transactionViolations]));
+        });
+
+        it('should add missingCategory violation to existing violations if they exist', () => {
+            transaction.category = undefined;
+            transactionViolations = [
+                {name: 'duplicatedTransaction', type: 'violation', userMessage: ''},
+                {name: 'receiptRequired', type: 'violation', userMessage: ''},
+            ];
+
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+            expect(result.value).toEqual(expect.arrayContaining([missingCategoryViolation, ...transactionViolations]));
+        });
+    });
+
+    describe('policyRequiresTags', () => {
+        beforeEach(() => {
+            policyRequiresTags = true;
+            policyTags = {Lunch: {enabled: true}, Dinner: {enabled: true}};
+            transaction.tag = 'Lunch';
+        });
+
+        it("shouldn't update the transactionViolations if the policy requires tags and the transaction has a tag from the policy", () => {
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+            expect(result.value).toEqual(transactionViolations);
+        });
+
+        it('should add a missingTag violation if none is provided and policy requires tags', () => {
+            transaction.tag = undefined;
+
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+            expect(result.value).toEqual(expect.arrayContaining([missingTagViolation]));
```

Comment:
I'm not sure if this is already being tested by this, but if not, let's add a test to check we're not adding a tag out of policy violation